bigframes/core/expression.py

```diff
@@ -0,0 +1,61 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import abc
+import dataclasses
+import itertools
+import typing
+
+import bigframes.operations
+
+
+@dataclasses.dataclass(frozen=True)
+class Expression(abc.ABC):
+    @property
+    def unbound_variables(self) -> typing.Tuple[str, ...]:
+        return ()
+
+
+@dataclasses.dataclass(frozen=True)
+class ScalarConstantExpression(Expression):
+    # TODO: Further constrain?
+    value: typing.Hashable
+
+
+@dataclasses.dataclass(frozen=True)
+class UnboundVariableExpression(Expression):
```

Comment:
What does unbound mean?

Reply:
Means its a "free" variable. For instance, in the expression "x + y + 3/x", the variables "x" and "y" are unbound.